src/main/scala/it/unibo/authsim/library/dsl/cryptography/encrypter/Cipher.scala

```diff
@@ -0,0 +1,112 @@
+package it.unibo.authsim.library.dsl.cryptography.encrypter
+
+import it.unibo.authsim.library.dsl.cryptography.algorithm.EncryptionAlgorithm
+import it.unibo.authsim.library.dsl.cryptography.algorithm.hash.HashFunction
+import it.unibo.authsim.library.dsl.cryptography.encrypter.asymmetric.key.KeyPair
+
+import java.security.spec.{AlgorithmParameterSpec, KeySpec}
+import java.util.Base64
+import javax.crypto.spec.{PBEKeySpec, PBEParameterSpec}
+import javax.crypto.{Cipher, SecretKey, SecretKeyFactory}
+
+/**
+ * Trait that represent a basic encryption
+ */
+trait Cipher:
+  def algorithm : EncryptionAlgorithm
+
+  /**
+   * Enumeration that specify in wich mode is to be used either encrypt or decrypt
+   */
+  protected enum EncryptionMode:
+    case Decryption, Encryption
+
+  /**
+   * Method used to encrypt a password using a generic secret
+   *
+   * @param password                password to be encrypted
+   * @param secret                  secret used to encrypt the password
+   * @tparam A                      generic parameter for the password
+   * @tparam B                      generic parameter for the secret
+   * @return                        a string representing the password encrypted
+   */
+  def encrypt[A, B](password: A, secret:B): String
+
+  /**
+   * Method used to decrypt a password using a generic secret
+   *
+   * @param encryptedPassword      encrypted password to be decrypted
+   * @param secret                 secret used to encrypt the password
+   * @tparam A                     generic parameter for the password
+   * @tparam B                     generic parameter for the secret
+   * @return                       a string representing the password decrypted
+   */
+  def decrypt[A, B](encryptedPassword: A, secret:B): String
+  
+/**
+ * Trait for Symmetric encrypter
+ */
+trait SymmetricEncrypter extends Cipher                                            //TODO non so se lasciarla o rimuoverla
```

Comment:
This trait could be useful to discriminate non-`AsymmetricEncrypter`, I'd keep it.

Reply:
oks thanks